stackrox/stackrox
# Import base paths from routePaths for Search links in UI

## Description

**Instead of** adding another special case to `getLink` function:

```diff
+} else if (item === 'POLICIES') {
+    link = `${link}/policy-management/policies`;
    }
    return `${link}/${lowerCase(item)}${id ? `/${id}` : ''}`;
```

Replace string manipulation of search categories with descriptor properties and lean into the future when requests from within containers replace global reducers and sagas.

1. Replace `mapping` and `filterOnMapping` objects with `searchCategoryDescriptorMap` object.

2. Replace `getLinks` function with `ViewLinks` component.

3. Replace `amendSearchOptions` and `onFilterLinkHandler` functions with `FilterLinks` component.

4. Replace `RelatedLink` elements with inline `Button` elements but factor out `NotAvailable` component.

5. Increase type safety for search service and simplify return types.

6. Adjust integration tests for `aria-label` instead of `id` attribute and title case instead of upper case.

### Residue

1. Find and replace any other links to deprecated policies path, and then delete `Redirect` element in `Body` component.
2. Evaluate pro and con to make separate request for each tab compared to front-end filtering by category.
3. Evaluate the amount of effort to replace `ReduxSearchInput` with `SearchFilterInput` element with PatternFly-ish style as in Policies and Violations.
4. Compare tabs to other precedents in UI.
5. Investigate why code implies that tabs which have 0 search results are disabled, but they are enabled.
    They display no results message which makes sense for **All** tab but might be misleading for the others: No results for your chosen filters Try changing the filter values.
6. Add integration tests for real requests.
7. What is `fieldToMatches` property of SearchResult object?

## Checklist

## Testing Performed

Test search and also containers which make REST requests instead of GraphQL queries for search options.

1. `yarn test` in ui/apps/platform

2. `yarn start` in ui

    * Visit /main/dashboard
        See GET /v1/search/metadata/options? request
        See corresponding search entries in `searchModifiers` and `searchSuggestions` under `globalSearch` in Redux
    
    * Visit /main/network
        See GET /v1/search/metadata/options?categories=DEPLOYMENTS request
        See corresponding options in search filter

    * Visit /main/policy-management/policies
        See GET /v1/search/metadata/options?categories=POLICIES request
        See corresponding options in search filter

    * Visit /main/violations
        See GET /v1/search/metadata/options?categories=ALERTS request
        See corresponding options in search filter
    
    * Click **Search** and then enter **Namespace: stackrox** in search filter input
        See GET /v1/search?query=Namespace%3Astackrox request
        See corresponding search results in `globalSearchResults` and `globalSearchCounts` under `globalSearch` in Redux

    * Click **Deployments** tab
        See GET /v1/search?query=Namespace%3Astackrox&categories=DEPLOYMENTS request
        See corresponding search results in `globalSearchResults` under `globalSearch` in Redux
        See `globalSearchCounts` does not change under `globalSearch` in Redux because it represents counts for all results

    * Click **Violations** tab interesting is 
        See GET /v1/search?query=Namespace%3Astackrox&categories=ALERTS request
        See corresponding search results in `globalSearchResults` under `globalSearch` in Redux
        See `globalSearchCounts` does not change under `globalSearch` in Redux because it represents counts for all results

    * Click **All** tab
        See GET /v1/search?query=Namespace%3Astackrox request
        See corresponding search results in `globalSearchResults` under `globalSearch` in Redux
        See `globalSearchCounts` does not change under `globalSearch` in Redux because it is deep equal to original request

3. `yarn cypress-open` in ui/apps/platform
    * search.test.js
